File: utils/settings_backup.go
package utils

import (
	"bytes"
	"encoding/json"
	"fmt"
	"runtime"
	"strings"
	"time"
)

const (
	SettingsBackupSchemaVersion = 2
	SettingsBackupKind          = "settings"
)

type SettingsBackupSourceApp struct {
	Name    string `json:"name"`
	Version string `json:"version"`
	Channel string `json:"channel"`
}

type SettingsBackupQuickInputSection struct {
	Pinned *[]string `json:"pinned,omitempty"`
	Recent *[]string `json:"recent,omitempty"`
}

func (s *SettingsBackupQuickInputSection) HasContent() bool {
	return s != nil && (s.Pinned != nil || s.Recent != nil)
}

type SettingsBackupServerPayload struct {
	AIAssistantStatus    *AIAssistantStatusConfig         `json:"aiAssistantStatus,omitempty"`
	Developer            *DeveloperConfig                 `json:"developer,omitempty"`
	PageTitle            *string                          `json:"pageTitle,omitempty"`
	DailyTip             *BackupDailyTipSettings          `json:"dailyTip,omitempty"`
	WebSessionQuickInput *SettingsBackupQuickInputSection `json:"webSessionQuickInput,omitempty"`
	Worktree             *WorktreeConfig                  `json:"worktree,omitempty"`
	TerminalShell        *SettingsBackupShellConfig       `json:"terminalShell,omitempty"`
	AuthAccess           *AuthAccessConfig                `json:"authAccess,omitempty"`
}

func (p *SettingsBackupServerPayload) HasContent() bool {
	return p != nil && (p.AIAssistantStatus != nil ||
		p.Developer != nil ||
		p.PageTitle != nil ||
		p.DailyTip != nil ||
		p.WebSessionQuickInput.HasContent() ||
		p.Worktree != nil ||
		p.TerminalShell != nil ||
		p.AuthAccess != nil)
}

type SettingsBackupShellConfig struct {
	Platform string `json:"platform"`
	Shell    string `json:"shell"`
}

type SettingsBackupClientPayload struct {
	Locale   *string         `json:"locale,omitempty"`
	Settings json.RawMessage `json:"settings,omitempty"`
}

func (p *SettingsBackupClientPayload) HasContent() bool {
	return p != nil && (p.Locale != nil || len(bytes.TrimSpace(p.Settings)) > 0)
}

type SettingsBackupPayload struct {
	Server *SettingsBackupServerPayload `json:"server,omitempty"`
	Client *SettingsBackupClientPayload `json:"client,omitempty"`
}

func (p SettingsBackupPayload) HasContent() bool {
	return p.Server.HasContent() || p.Client.HasContent()
}

type SettingsBackupExportOptions struct {
	IncludeServer           bool   `json:"includeServer,omitempty"`
	IncludeClient           bool   `json:"includeClient,omitempty"`
	IncludeSecurityAccess   bool   `json:"includeSecurityAccess,omitempty"`
	IncludeQuickInputRecent bool   `json:"includeQuickInputRecent,omitempty"`
	FileNameRule            string `json:"fileNameRule,omitempty"`
	IncludeMetadata         bool   `json:"includeMetadata,omitempty"`
}

type SettingsBackupMeta struct {
	Description   string                       `json:"description,omitempty"`
	ExportOptions *SettingsBackupExportOptions `json:"exportOptions,omitempty"`
}

type SettingsBackupFile struct {
	BackupSchemaVersion int                     `json:"backupSchemaVersion"`
	BackupKind          string                  `json:"backupKind"`
	CreatedAt           *time.Time              `json:"createdAt,omitempty"`
	SourceApp           SettingsBackupSourceApp `json:"sourceApp"`
	Meta                *SettingsBackupMeta     `json:"meta,omitempty"`
	Payload             SettingsBackupPayload   `json:"payload"`
}

type SettingsBackupPreviewIssue struct {
	Code    string `json:"code"`
	Level   string `json:"level"`
	Message string `json:"message"`
}

type SettingsBackupPreviewSection struct {
	Key          string   `json:"key"`
	Label        string   `json:"label"`
	Action       string   `json:"action"`
	Target       string   `json:"target"`
	ChangedKeys  []string `json:"changedKeys,omitempty"`
	WarningCodes []string `json:"warningCodes,omitempty"`
}

type SettingsBackupPreviewResult struct {
	BackupSchemaVersion int                            `json:"backupSchemaVersion"`
	BackupKind          string                         `json:"backupKind"`
	SourceApp           SettingsBackupSourceApp        `json:"sourceApp"`
	CurrentApp          SettingsBackupSourceApp        `json:"currentApp"`
	CanImport           bool                           `json:"canImport"`
	Errors              []SettingsBackupPreviewIssue   `json:"errors"`
	Warnings            []SettingsBackupPreviewIssue   `json:"warnings"`
	Sections            []SettingsBackupPreviewSection `json:"sections"`
	Migrated            bool                           `json:"migrated"`
}

func BuildSettingsBackupServerPayload(cfg *AppConfig) SettingsBackupServerPayload {
	currentShell := strings.TrimSpace(CurrentPlatformShell(cfg.Terminal.Shell))
	quickInput := NormalizeWebSessionQuickInputConfig(cfg.UI.WebSessionQuickInput)
	aiStatus := cfg.Terminal.AIAssistantStatus
	developer := NormalizeDeveloperConfig(cfg.Developer)
	dailyTip := BackupDailyTipSettings{Enabled: cfg.UI.DailyTipEnabled}
	worktree := cfg.Worktree
	terminalShell := SettingsBackupShellConfig{
		Platform: runtime.GOOS,
		Shell:    currentShell,
	}
	authAccess := SanitizeAuthAccessConfig(AuthAccessConfigFromAuthConfig(cfg.Auth))

	return SettingsBackupServerPayload{
		AIAssistantStatus: ptrValue(aiStatus),
		Developer:         ptrValue(developer),
		PageTitle:         ptrValue(cfg.UI.PageTitle),
		DailyTip:          ptrValue(dailyTip),
		WebSessionQuickInput: &SettingsBackupQuickInputSection{
			Pinned: ptrStringSlice(quickInput.Pinned),
			Recent: ptrStringSlice(quickInput.Recent),
		},
		Worktree:      ptrValue(worktree),
		TerminalShell: ptrValue(terminalShell),
		AuthAccess:    ptrValue(authAccess),
	}
}

type BackupDailyTipSettings struct {
	Enabled bool `json:"enabled"`
}

func CurrentPlatformShell(cfg TerminalShellConfig) string {
	switch runtime.GOOS {
	case "windows":
		return cfg.Windows
	case "darwin":
		return cfg.Darwin
	default:
		return cfg.Linux
	}
}

func ApplyCurrentPlatformShell(cfg *TerminalShellConfig, shell string) {
	if cfg == nil {
		return
	}
	switch runtime.GOOS {
	case "windows":
		cfg.Windows = shell
	case "darwin":
		cfg.Darwin = shell
	default:
		cfg.Linux = shell
	}
}

func MigrateSettingsBackupFile(input SettingsBackupFile) (SettingsBackupFile, bool, error) {
	switch input.BackupSchemaVersion {
	case SettingsBackupSchemaVersion:
		return input, false, nil
	case 0:
		return SettingsBackupFile{}, false, fmt.Errorf("backup schema version is required")
	default:
		return SettingsBackupFile{}, false, fmt.Errorf("unsupported backup schema version: %d", input.BackupSchemaVersion)
	}
}

func ValidateSettingsBackupFileShape(input SettingsBackupFile) error {
	if input.BackupSchemaVersion <= 0 {
		return fmt.Errorf("backupSchemaVersion is required")
	}
	if strings.TrimSpace(input.BackupKind) == "" {
		return fmt.Errorf("backupKind is required")
	}
	if input.BackupKind != SettingsBackupKind {
		return fmt.Errorf("unsupported backup kind: %s", input.BackupKind)
	}
	if !input.Payload.HasContent() {
		return fmt.Errorf("backup payload is empty")
	}
	return nil
}

func NormalizeSettingsBackupQuickInputSection(
	section *SettingsBackupQuickInputSection,
) *SettingsBackupQuickInputSection {
	if section == nil {
		return nil
	}
	normalized := &SettingsBackupQuickInputSection{}
	if section.Pinned != nil {
		pinned := NormalizeWebSessionQuickInputConfig(WebSessionQuickInputConfig{
			Pinned: append([]string(nil), (*section.Pinned)...),
		}).Pinned
		normalized.Pinned = ptrStringSlice(pinned)
	}
	if section.Recent != nil {
		recent := NormalizeWebSessionQuickInputConfig(WebSessionQuickInputConfig{
			Recent: append([]string(nil), (*section.Recent)...),
		}).Recent
		normalized.Recent = ptrStringSlice(recent)
	}
	if !normalized.HasContent() {
		return nil
	}
	return normalized
}

func ptrValue[T any](value T) *T {
	return &value
}

func ptrStringSlice(values []string) *[]string {
	items := append([]string(nil), values...)
	return &items
}

File: ui/src/utils/settingsBackup.ts
import { extractItem } from '@/api/response';
import { http } from '@/api/http';
import type { AIAssistantStatusConfig, DeveloperConfig, WorktreeConfig } from '@/types/models';
import type {
  EditorSettings,
  ShortcutSettings,
  TerminalFontSettings,
  TerminalQuickAction,
  ThemeSettings,
  WebSessionQuickInputSettings,
  WebSessionAutoContinuePreset,
  WebSessionAutoContinueScope,
  WebSessionStreamingMarkdownThrottleMode,
  FollowSystemThemeSetting,
} from '@/stores/settings';
import type { AuthAccessConfig } from '@/stores/auth';
import type { TerminalRenderMode } from '@/constants/terminalRenderMode';
import type { TerminalConnectionPolicy } from '@/constants/terminalConnectionPolicy';
import type { WebSessionActivityDisplayMode } from '@/constants/webSessionActivityDisplayMode';
import { useAppStore } from '@/stores/app';

export const SETTINGS_BACKUP_SCHEMA_VERSION = 2;
export const SETTINGS_BACKUP_KIND = 'settings';
export const SETTINGS_STORAGE_KEY = 'general_settings';
export const APP_LOCALE_STORAGE_KEY = 'app-locale';

export type SettingsBackupFileNameRule =
  | 'app-version-date'
  | 'app-version-datetime'
  | 'channel-app-version-datetime';

export type SettingsBackupImportMode = 'preview-only' | 'preview-confirm' | 'direct-import';
export type SettingsBackupVersionWarningMode = 'standard' | 'strict';

export interface SettingsBackupShellConfig {
  platform: string;
  shell: string;
}

export interface SettingsBackupQuickInputSection {
  pinned?: string[];
  recent?: string[];
}

export interface SettingsBackupServerPayload {
  aiAssistantStatus?: AIAssistantStatusConfig;
  developer?: DeveloperConfig;
  pageTitle?: string;
  dailyTip?: {
    enabled: boolean;
  };
  webSessionQuickInput?: SettingsBackupQuickInputSection;
  worktree?: WorktreeConfig;
  terminalShell?: SettingsBackupShellConfig;
  authAccess?: AuthAccessConfig;
}

export interface SettingsBackupClientSettings {
  version: number;
  theme: ThemeSettings;
  currentPresetId: string;
  followSystemTheme: FollowSystemThemeSetting;
  customTheme: ThemeSettings | null;
  recentProjectsLimit: number;
  maxTerminalsPerProject: number;
  panelShortcuts: ShortcutSettings;
  webSessionQuickInput?: Partial<WebSessionQuickInputSettings>;
  webSessionQuickInputDirectSend: boolean;
  terminalQuickActions: TerminalQuickAction[];
  editor: EditorSettings;
  confirmBeforeTerminalClose: boolean;
  showWebSessionReasoning: boolean;
  webSessionActivityDisplayMode: WebSessionActivityDisplayMode;
  webSessionAutoContinueScope: WebSessionAutoContinueScope;
  webSessionAutoContinuePreset: WebSessionAutoContinuePreset;
  webSessionAutoContinueMaxAttempts: number;
  webSessionAutoRetryDispatchPendingOnFailure: boolean;
  webSessionStreamingMarkdownThrottleMode: WebSessionStreamingMarkdownThrottleMode;
  webSessionStreamingMarkdownThrottleCustomMs: number;
  terminalThemeId: string;
  terminalFont: TerminalFontSettings;
  terminalWebGLRenderer: 'auto' | 'force' | 'disable';
  defaultTerminalRenderMode: TerminalRenderMode;
  defaultTerminalSnapshotIntervalMs: number | null;
  defaultTerminalSnapshotZlibCompression: boolean;
  terminalConnectionPolicy: TerminalConnectionPolicy;
  inactiveTerminalSnapshotIntervalMs: number;
}

export interface SettingsBackupClientPayload {
  locale?: string;
  settings?: SettingsBackupClientSettings;
}

export interface SettingsBackupSourceApp {
  name: string;
  version: string;
  channel: string;
}

export interface SettingsBackupExportOptions {
  includeServer: boolean;
  includeClient: boolean;
  includeSecurityAccess: boolean;
  includeQuickInputRecent: boolean;
  fileNameRule: SettingsBackupFileNameRule;
  includeMetadata: boolean;
}

export interface SettingsBackupMeta {
  description?: string;
  exportOptions?: SettingsBackupExportOptions;
}

export interface SettingsBackupFile {
  backupSchemaVersion: number;
  backupKind: string;
  createdAt?: string;
  sourceApp: SettingsBackupSourceApp;
  meta?: SettingsBackupMeta;
  payload: {
    server?: SettingsBackupServerPayload;
    client?: SettingsBackupClientPayload;
  };
}

export interface SettingsBackupPreviewIssue {
  code: string;
  level: 'warning' | 'error';
  message: string;
}

export interface SettingsBackupPreviewSection {
  key: string;
  label: string;
  action: string;
  target: 'server' | 'client';
  changedKeys?: string[];
  warningCodes?: string[];
}

export interface SettingsBackupPreviewResult {
  backupSchemaVersion: number;
  backupKind: string;
  sourceApp: SettingsBackupSourceApp;
  currentApp: SettingsBackupSourceApp;
  canImport: boolean;
  errors: SettingsBackupPreviewIssue[] | null;
  warnings: SettingsBackupPreviewIssue[] | null;
  sections: SettingsBackupPreviewSection[] | null;
  migrated: boolean;
}

type ItemResponse<T> = {
  item?: T;
};

export async function exportServerSettingsBackup(): Promise<SettingsBackupFile> {
  const response = await http
    .Get<ItemResponse<SettingsBackupFile>>('/system/settings-backup/export')
    .send();
  const item = extractItem<SettingsBackupFile>(response);
  if (!item) {
    throw new Error('failed to export settings backup');
  }
  return item;
}

export async function previewSettingsBackup(
  backup: SettingsBackupFile
): Promise<SettingsBackupPreviewResult> {
  const response = await http
    .Post<ItemResponse<SettingsBackupPreviewResult>>('/system/settings-backup/preview', backup)
    .send();
  const item = extractItem<SettingsBackupPreviewResult>(response);
  if (!item) {
    throw new Error('failed to preview settings backup');
  }
  return item;
}

export async function importSettingsBackup(
  backup: SettingsBackupFile
): Promise<SettingsBackupPreviewResult> {
  const response = await http
    .Post<ItemResponse<SettingsBackupPreviewResult>>('/system/settings-backup/import', backup)
    .send();
  const item = extractItem<SettingsBackupPreviewResult>(response);
  if (!item) {
    throw new Error('failed to import settings backup');
  }
  return item;
}

export function buildSettingsBackupFile(options: {
  serverBackup: SettingsBackupFile;
  clientPayload: SettingsBackupClientPayload;
  exportOptions: SettingsBackupExportOptions;
}): SettingsBackupFile {
  const appStore = useAppStore();
  const filteredServer = filterServerPayloadForExport(
    options.serverBackup.payload.server,
    options.exportOptions
  );
  const filteredClient = filterClientPayloadForExport(options.clientPayload, options.exportOptions);
  const createdAt = options.exportOptions.includeMetadata ? new Date().toISOString() : undefined;
  const meta = options.exportOptions.includeMetadata
    ? {
        description: buildSettingsBackupDescription(options.exportOptions),
        exportOptions: { ...options.exportOptions },
      }
    : undefined;

  return {
    backupSchemaVersion: SETTINGS_BACKUP_SCHEMA_VERSION,
    backupKind: SETTINGS_BACKUP_KIND,
    createdAt,
    sourceApp: {
      name: appStore.appInfo.name,
      version: appStore.appInfo.version,
      channel: appStore.appInfo.channel,
    },
    meta,
    payload: {
      ...(filteredServer ? { server: filteredServer } : {}),
      ...(filteredClient ? { client: filteredClient } : {}),
    },
  };
}

export function hasSettingsBackupContent(backup: SettingsBackupFile | null | undefined) {
  if (!backup) {
    return false;
  }
  return Boolean(
    hasServerPayloadContent(backup.payload.server) || hasClientPayloadContent(backup.payload.client)
  );
}

export function parseSettingsBackupJSON(text: string): SettingsBackupFile {
  const parsed = JSON.parse(text) as SettingsBackupFile;
  if (!parsed || typeof parsed !== 'object') {
    throw new Error('invalid backup JSON');
  }
  return parsed;
}

export function downloadSettingsBackupFile(backup: SettingsBackupFile, fileName: string) {
  const blob = new Blob([JSON.stringify(backup, null, 2)], { type: 'application/json' });
  const objectURL = URL.createObjectURL(blob);
  const anchor = document.createElement('a');
  anchor.href = objectURL;
  anchor.download = fileName;
  anchor.click();
  window.setTimeout(() => URL.revokeObjectURL(objectURL), 0);
}

export function formatSettingsBackupFileName(options: {
  appInfo?: Partial<SettingsBackupSourceApp> | null;
  rule: SettingsBackupFileNameRule;
  createdAt?: string | null;
}) {
  const version = sanitizeFileSegment(options.appInfo?.version || '0.0.0');
  const channel = sanitizeFileSegment(options.appInfo?.channel || 'unknown');
  const createdAt = options.createdAt ? new Date(options.createdAt) : new Date();
  const date = `${createdAt.getUTCFullYear()}${String(createdAt.getUTCMonth() + 1).padStart(2, '0')}${String(createdAt.getUTCDate()).padStart(2, '0')}`;
  const datetime = `${date}-${String(createdAt.getUTCHours()).padStart(2, '0')}${String(createdAt.getUTCMinutes()).padStart(2, '0')}${String(createdAt.getUTCSeconds()).padStart(2, '0')}`;

  switch (options.rule) {
    case 'app-version-date':
      return `codekanban-settings-v${version}-${date}.json`;
    case 'channel-app-version-datetime':
      return `codekanban-settings-${channel}-v${version}-${datetime}.json`;
    default:
      return `codekanban-settings-v${version}-${datetime}.json`;
  }
}

export function buildImportableSettingsBackup(
  backup: SettingsBackupFile,
  selectedSectionKeys: string[]
): SettingsBackupFile {
  const cloned = cloneBackup(backup);
  const selectedKeys = new Set(selectedSectionKeys);
  const server = cloned.payload.server;
  const client = cloned.payload.client;

  if (server) {
    if (!selectedKeys.has('server.aiAssistantStatus')) {
      delete server.aiAssistantStatus;
    }
    if (!selectedKeys.has('server.developer')) {
      delete server.developer;
    }
    if (!selectedKeys.has('server.pageTitle')) {
      delete server.pageTitle;
    }
    if (!selectedKeys.has('server.dailyTip')) {
      delete server.dailyTip;
    }
    if (server.webSessionQuickInput) {
      if (!selectedKeys.has('server.webSessionQuickInput.pinned')) {
        delete server.webSessionQuickInput.pinned;
      }
      if (!selectedKeys.has('server.webSessionQuickInput.recent')) {
        delete server.webSessionQuickInput.recent;
      }
      if (!hasQuickInputSectionContent(server.webSessionQuickInput)) {
        delete server.webSessionQuickInput;
      }
    }
    if (!selectedKeys.has('server.worktree')) {
      delete server.worktree;
    }
    if (!selectedKeys.has('server.terminalShell')) {
      delete server.terminalShell;
    }
    if (!selectedKeys.has('server.authAccess')) {
      delete server.authAccess;
    }
    if (!hasServerPayloadContent(server)) {
      delete cloned.payload.server;
    }
  }

  if (client) {
    if (!selectedKeys.has('client.locale')) {
      delete client.locale;
    }
    if (!selectedKeys.has('client.settings')) {
      delete client.settings;
    } else if (client.settings?.webSessionQuickInput) {
      if (!selectedKeys.has('server.webSessionQuickInput.pinned')) {
        delete client.settings.webSessionQuickInput.pinned;
      }
      if (!selectedKeys.has('server.webSessionQuickInput.recent')) {
        delete client.settings.webSessionQuickInput.recent;
      }
      if (!hasQuickInputSectionContent(client.settings.webSessionQuickInput)) {
        delete client.settings.webSessionQuickInput;
      }
    }
    if (!hasClientPayloadContent(client)) {
      delete cloned.payload.client;
    }
  }

  return cloned;
}

function filterServerPayloadForExport(
  payload: SettingsBackupServerPayload | undefined,
  options: SettingsBackupExportOptions
) {
  if (!payload || !options.includeServer) {
    return undefined;
  }

  const cloned = cloneBackup({ payload: { server: payload } } as SettingsBackupFile).payload.server;
  if (!cloned) {
    return undefined;
  }

  if (!options.includeSecurityAccess) {
    delete cloned.authAccess;
  }
  if (cloned.webSessionQuickInput && !options.includeQuickInputRecent) {
    delete cloned.webSessionQuickInput.recent;
    if (!hasQuickInputSectionContent(cloned.webSessionQuickInput)) {
      delete cloned.webSessionQuickInput;
    }
  }

  return hasServerPayloadContent(cloned) ? cloned : undefined;
}

function filterClientPayloadForExport(
  payload: SettingsBackupClientPayload | undefined,
  options: SettingsBackupExportOptions
) {
  if (!payload || !options.includeClient) {
    return undefined;
  }

  const cloned = cloneBackup({ payload: { client: payload } } as SettingsBackupFile).payload.client;
  if (!cloned) {
    return undefined;
  }

  if (cloned.settings?.webSessionQuickInput && !options.includeQuickInputRecent) {
    delete cloned.settings.webSessionQuickInput.recent;
    if (!hasQuickInputSectionContent(cloned.settings.webSessionQuickInput)) {
      delete cloned.settings.webSessionQuickInput;
    }
  }

  return hasClientPayloadContent(cloned) ? cloned : undefined;
}

function buildSettingsBackupDescription(options: SettingsBackupExportOptions) {
  return `Settings backup: server=${options.includeServer ? 'yes' : 'no'}, client=${options.includeClient ? 'yes' : 'no'}, securityAccess=${options.includeSecurityAccess ? 'yes' : 'no'}, quickInputRecent=${options.includeQuickInputRecent ? 'yes' : 'no'}.`;
}

function hasServerPayloadContent(payload: SettingsBackupServerPayload | undefined) {
  if (!payload) {
    return false;
  }
  return Boolean(
    payload.aiAssistantStatus ||
      payload.developer ||
      typeof payload.pageTitle === 'string' ||
      payload.dailyTip ||
      hasQuickInputSectionContent(payload.webSessionQuickInput) ||
      payload.worktree ||
      payload.terminalShell ||
      payload.authAccess
  );
}

function hasClientPayloadContent(payload: SettingsBackupClientPayload | undefined) {
  if (!payload) {
    return false;
  }
  return Boolean(payload.locale || payload.settings);
}

function hasQuickInputSectionContent(
  value: Partial<WebSessionQuickInputSettings> | SettingsBackupQuickInputSection | undefined
) {
  if (!value) {
    return false;
  }
  return 'pinned' in value || 'recent' in value;
}

function sanitizeFileSegment(value: string) {
  return (
    String(value || '')
      .trim()
      .replace(/[^a-zA-Z0-9._-]+/g, '-')
      .replace(/-+/g, '-')
      .replace(/^-|-$/g, '') || 'unknown'
  );
}

function cloneBackup<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}
